File: Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        
        ans = []
        num2letter = ['abc','def','ghi','jkl','mno','pqrs','tuv','wxyz']
        
        for number in digits:
            pre_ans = []
            
            # append new letters to the previous letter sequences -> like expending a tree
            if not ans:
                for letter in num2letter[int(number) - 2]:
                    pre_ans.append(letter)
            else:
                for letter in num2letter[int(number) - 2]:
                    for exist_letter in ans:
                        pre_ans.append(exist_letter + letter)
            
            ans = pre_ans.copy()
        
        return ans

File: C++/Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> ans;
        if(digits.length() == 0)
            return ans;
        
        ans.push_back("");
        string str[8] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        
        // renew the combination result for every iteration
        // append new character to the tail
        for(int i = 0; i < digits.length(); i++){
            vector<string> new_ans;
            
            string alternative = str[digits[i] - '0' - 2];
            for(int j = 0; j < alternative.length(); j++)
                for(int k = 0; k < ans.size(); k++)
                    new_ans.push_back(ans[k] + alternative[j]);
            
            ans = new_ans;
        }
        
        return ans;
    }
};